### DISCUSSION TITLE
Reference another project (tool) but not make it available in referencing project

### DISCUSSION THREAD

**Discussion by vladd** [dotnet/msbuild#10440] [Q&A]
I've got the following setup:

```
Solution
    ToolProject
    LibraryProject
    MainAppProject
```

LibraryProject is a .NET library referenced by .NET MainAppProject.
ToolProject is a .NET project too, it compiles to an executable which is sometimes run by the LibraryProject (I cannot link against ToolProject because it needs a different bitness).

Now, I want the output of ToolProject be available for the MainApp (as a transitive dependency of LibraryProject).

For it, I tried using
```
<ProjectReference Include="..\ToolProject.csproj">
    <ReferenceOutputAssembly>False</ReferenceOutputAssembly>
</ProjectReference>
```
in LibraryProject.csproj, but this way only Tool.exe is copied to the target folder, not the other files.

I tried using
```
<ProjectReference Include="..\ToolProject.csproj">
    <ReferenceOutputAssembly>True</ReferenceOutputAssembly>
    <OutputItemType>Content</OutputItemType>
    <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
</ProjectReference>
```

This achieves the job, but it has an unfortunate effect that the classes from ToolProject become available in the LibraryProject, which is very misleading and leads to conflicts.

The question: is it possible to reference the ToolProject so that it will be used as a dependency from the project point of view (compile order, copying to the output, etc.), but will not be visible in the LibraryProject's source code?

**Comment by rainersigwald** [comment]
This isn't very easy right now. The main tracking bug is https://github.com/dotnet/runtime/issues/53834.

**Reply by vladd** [reply]
I've looked at the referenced issue, but it seems to be about publishing and self-contained apps, right? I don't directly see how could it help in my case, although the issue seems to be connected. My problem seems to be more like plugin distribution, where the main code depends on plugin but doesn't want to reference it in C# sense.

**Reply by vladd** [reply]
Maybe let's look at the problem from another angle. Suppose that I need to copy the Tool executable and its dependencies into the subfolder Tools near the MainApp, but still want it to be a dependency so that changes in Tool get copied whenever MainApp is compiled. Is this achievable through the project system?

**Reply by JanKrivanek** [reply]
It sounds like you want `DisableTransitiveProjectReferences` https://github.com/dotnet/msbuild/blob/main/documentation/wiki/Controlling-Dependencies-Behavior.md#access-to-transitive-project-references - is that what you are looking for?

**Reply by vladd** [reply]
@JanKrivanek This seems pretty close. With `DisableTransitiveProjectReferences` I can disallow references to the downstream projects (in my case, to the libraries which are used by ToolProject). But what I actually need is disabling references to the ToolProject itself, too.

**Reply by JanKrivanek** [reply]
@vladd - If I understand you correctly - you are interested only in copying the build outputs, without exposing the reference public surface to other project(s) in the build - is that correct? There is no short-cut reference metadata support for those requirements. You'd need to write a custom target yourself (that copies the content of the OutputPath after build to the directory of interest). But perhaps the [Artifacts output layout](https://learn.microsoft.com/en-us/dotnet/core/sdk/artifacts-output) might help you achieve what you are looking for?

**Reply by vladd** [reply]
@JanKrivanek 
> you are interested only in copying the build outputs, without exposing the reference public surface to other project(s) in the build - is that correct?

Exactly, this is my goal.

>  You'd need to write a custom target yourself (that copies the content of the OutputPath after build to the directory of interest).

Well, this is perhaps going to be tricky, as I would need to specify the build order dependency somehow, right?

**Reply by JanKrivanek** [reply]
You should still be able to use the `ReferenceOutputAssembly` to guarantee the build order. The deps copying would however be on you.

But please try to first have a look on [Artifacts output layout](https://learn.microsoft.com/en-us/dotnet/core/sdk/artifacts-output) to see if it can help you achive your goal.

**Reply by Nirmal4G** [reply]
See, that's the thing..., in most Plug-in or tools based project structure, we don't need the reference but want the build order and the copying of dependencies.

I struggled with this problem quite a lot and IMHO, setting `ReferenceOutputAssembly=False` should ensure build order. The `ProjectReference` should also respect `Pack`, `Publish` and `CopyToOutputDirectory`.

**Comment by Nirmal4G** [comment]
This isn't possible with `ProjectReference` right now but you can specify project dependencies within the solution through the Solution Properties dialog in Visual Studio. Remove the `ProjectReference` before setting the Project dependencies in the solution.

After that, MSBuild/VisualStudio will always build in the expected order without referencing the tool project.

**Reply by vladd** [reply]
Unfortunately, build order seems to be not enough, because I need the project output to be copied to the referencing project’s output folder.

**Reply by Nirmal4G** [reply]
Build order ensures that the dependencies will always be built first and then you can gather the build outputs and copy them manually into the referencing project.

Just like here: CommunityToolkit/dotnet#220

See the `Dynamically Pack...` commit. Especially the `...GetBuildOutputs.targets` file which shows how to gather the outputs and the `...Packaging.targets` file which shows how to use it.

**Reply by vladd** [reply]
Well, it's not that easy, since the referencing project may be a library as well, so we need it to be a transitive dependency for copying. Sorry for being unclear in my previous answer.

**Reply by vladd** [reply]
The customary `<ProjectReference>` means from user's point of view 2 things:
* the public classes from the referenced project are available in the referencing project
* the output files of the referenced project appear in the output folder of the executable project which is
   * either referencing project itself if it's executable;
   * or, if the referencing project is a library, in every executable project referencing (maybe indirectly) this library.

My need is to remove the first bullet while keeping the second (because the tool will be started by the referencing library as a process). (I need the referenced project to be a separate tool, because it has different bitness.)

**Reply by Nirmal4G** [reply]
Exactly what I was trying to solve in my own projects. I don't want to reference the files in Code but want them in output/publish directory.

In my own projects, I never went beyond 2 levels of transitive references, so, I can't confidently say that this might work beyond that but it's a starting point.

You can look at the [binlog](https://msbuildlog.com) to see where and how the transitive references flow and then hook these targets at appropriate places and add the transitive Items either in `BuildOutputPerTarget` or `BuildOutput` items inside the `GetBuildOutput...` targets. This will get you the entire output including transitive references.
